illumio_connector.py

```diff
@@ -152,6 +151,286 @@ def _handle_get_traffic_analysis(self, param):
+
+        protocol = param["protocol"].lower()
+        if protocol not in PROTOCOL_LIST:
+            return action_result.set_status(
+                phantom.APP_ERROR, ILLUMIO_INVALID_PROTOCOL_MSG
+            )
+
+        service_name = param["name"]
+        virtual_service = None
+
+        ret_val = self.connect_pce(action_result)
+        if phantom.is_fail(ret_val):
+            return action_result.get_status()
+
+        try:
+            service_list = self._pce.virtual_services.get(
+                params={"name": "{}".format(service_name)}
+            )
+            for service in service_list:
+                if service_name == service.name:
+                    virtual_service = service
+                    action_result.set_status(
+                        phantom.APP_SUCCESS,
+                        "Found existing virtual service with name {}".format(
+                            service_name
+                        ),
+                    )
+                    break
+
+            if not virtual_service:
+                virtual_service = illumio.VirtualService(
+                    name=service_name,
+                    service_ports=[illumio.ServicePort(port=port, proto=protocol)],
+                )
+                virtual_service = self._pce.virtual_services.create(virtual_service)
+                action_result.set_status(
+                    phantom.APP_SUCCESS, "Successfully created virtual service"
+                )
+
+        except IllumioException as e:
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                "Encountered error creating virtual service: {}".format(e),
+            )
+
+        result = self.convert_object_to_json(virtual_service, action_result)
+        action_result.add_data(result)
+        return action_result.get_status()
+
+    def _handle_provision_objects(self, param):
+        hrefs = param["hrefs"]
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+
+        ret_val = self.connect_pce(action_result)
+        if phantom.is_fail(ret_val):
+            return action_result.get_status()
+
+        try:
+            hrefs_list = self.handle_comma_seperated_string(hrefs)
+
+            provisioned_virtual_service_obj = self._pce.provision_policy_changes(
+                change_description="Phantom object provisioning.",
+                hrefs=hrefs_list,
+            )
+
+        except IllumioException as e:
+            return action_result.set_status(
+                phantom.APP_ERROR,
+                "Encountered error provisioning object: {}".format(e),
+            )
+
+        result = self.convert_object_to_json(
+            provisioned_virtual_service_obj, action_result
+        )
+        action_result.add_data(result)
+        return action_result.set_status(
+            phantom.APP_SUCCESS, "Successfully provisioned object"
+        )
+
+    def _handle_get_ip_lists(self, param):
```

Comment:
This function should accept more parameters than just the name, and should return all IP lists found.

Reply:
While TDD creation it was discussed that, this action should fetch IP List according to name instead of name=Any(0.0.0.0/0 and ::/0). Hence, we added only name as input parameter and made it more generic.

Reply:
The outcome of that discussion was to change the original get_ip_list(name) function to the more generic get_ip_lists(params), and to have a get_ip_list(href) function.

It doesn't make sense to have a function called get_ip_lists that returns one IP list instance by name.

Reply:
We have added name, fqdn, ip address and description filter parameters in get_ip_lists actions as these filter parameters were present on UI